### PR TITLE
Zen 563 filter modal

### PR DESCRIPTION
**Pull request by subaya710**
**Ticket**: https://jira.simpleviewtools.com/browse/ZEN-563

## Context

When certain labels do not exist , the option to filter by that label should not appear. Also , when there is nothing that can actually be filtered , the Filter button is unnecessary

## Goal

* Ensure that when for all sessions 'isWaitingListAvailable' option is set to false , the 'Waiting List' and 'On waiting List' options do not appear in the filter labels
* Ensure that when for all sessions 'allowBooking' option is set to false , the 'Available' and 'Selected' options do not appear in the filter labels
* Ensure that when there are no custom labels AND there are no sessions with a true 'isWaitingListAvailable' or 'allowBooking' option , the filter option/button is not present since there is nothing to filter by

## Updates

* First part of updates was code to ensure 'Available' and 'Selected' appeared / disappeared based on the incoming data allowBooking and the second part was to ensure filter option disappeared when there were no labels
* *tap-events-force/src/hooks/sessions/useAllowBookingActive.js*
  * Hook to update the store object based on AllowBookingActive
* *tap-events-force/src/components/modals/filter.js*
  * Apply logic appropriately to show 'Available' and 'Selected' OR 'Waiting List' and 'On waiting List'  OR both OR none based on AllowBookingActive and WaitingListActive
* *tap-events-force/src/actions/session/filters/applySessionFilters.js*
  * Changes to apply session filters by also calling AllowBookingActive
* *tap-events-force/src/actions/session/booking/setAllowBookingActive.js*
  * Test the boolean value of the allowBooking field at the root of the session data object 
* *tap-events-force/src/components/sessionsHeader/sessionsHeader.js*
  * Changes to display the 'Right Component' which is the parent of the Filter button , only when some labels to filter on are present

## Testing
*Setup
- Run the command => keg evf pack run evf:zen-563-filtermodal
- Go to http://evf-ZEN-563-FilterModal.local.keghub.io

Note 1: Docker image : docker pull ghcr.io/simpleviewinc/evf:zen-563-filtermodal
Note 2: Images for testing attached at the end

* Test 1 : To confirm if Waiting List functionality is working
  * Launch events force from the docker image above
  * Click on Filter link on top right and confirm that 'Waiting List' and 'On Waiting List' labels are present
  * Click on the 'Test Data (JSON)' button at the top left
  * Search 'Sessions' section , and within that find every session where isWaitingListAvailable: true and set it to false, Apply changes
  * Click the Filter link again and confirm that 'Waiting List' and 'On Waiting List'  are no longer present

* Test 2 : To confirm if Select/Available functionality is working
  * Launch events force from the docker image above
  * Click on Filter link on top right and confirm that 'Available' and 'Selected' labels are present
  * Click on the 'Test Data (JSON)' button at the top left
  * Search 'Sessions' section , and within that find every session where allowBooking: true and set it to false, Apply changes
  * Click the Filter link again and confirm that 'Available' and 'Selected' are no longer present

* Test 3 : To confirm if Filter link disappear functionality is working
  * Launch events force from the docker image above
  * Click on Filter link on top right and confirm that all labels are present. Custom labels at the top section and default lables at the bottom section of the modal
  * Click on the 'Test Data (JSON)' button at the top left
  * Search 'Sessions' section , and within that find every session where allowBooking: true and set it to false, Apply changes
  * Search 'Sessions' section , and within that find every session where isWaitingListAvailable: true and set it to false, Apply changes
  * Find the labels section and comment it out entirely
  * These 3 steps means there are no labels that can be valid filters. Therefore there is no necessity for the filter link
  * Ensure Filter link is no longer present

Attached below are images taken from my testing and that will assist in yours
*Image 1: Starting with all labels / filter options present
ScreenShot-1

*Image 2: When there are no isWaitingListAvailable: true  but atleast one allowBooking: true
ScreenShot-2

*Image 3: When there are no allowBooking: true but atleast one isWaitingListAvailable: true 
ScreenShot-3

*Image 4: When there are no isWaitingListAvailable: true OR allowBooking: true
ScreenShot-4

*Image 5: When there are no custom labels but default labels are present because of atleast one allowBooking: true and isWaitingListAvailable: true
ScreenShot-5

*Image 6: When there are no custom labels OR default labels present 
ScreenShot-6

### CODE REVIEW

File: src/components/modals/filter.js [hunk 1]
```diff
@@ -193,21 +193,24 @@ const LabelButtons = React.memo(
  */
 const useStateLabels = bookingStates => {
   const waitingListActive = useWaitingListActive()
+  const allowBookingActive = useAllowBookingActive()
```
Comment (mpcarolin) [inline src/components/modals/filter.js:196]: The "active" wording here seems out of place to me. I think "allowBooking" and 'useAllowBooking" would be sufficient.

File: src/components/modals/filter.js [hunk 2]
```diff
@@ -193,21 +193,24 @@ const LabelButtons = React.memo(
...
+    const { WAITING_LIST, ON_WAITING_LIST, SELECT, SELECTED } = SESSION_BOOKING_STATES
     return reduceObj(
       bookingStates,
       (key, value, labels) => {
-        ;(waitingListActive ||
-          (value !== WAITING_LIST && value !== ON_WAITING_LIST)) &&
+        ;((waitingListActive &&
+          (value == WAITING_LIST || value == ON_WAITING_LIST)) ||
+          (allowBookingActive &&
+          (value == SELECT || value == SELECTED))) &&
```
Comment (mpcarolin) [inline src/components/modals/filter.js:206]: With the addition of this new Boolean condition, I think this whole Boolean expression is becoming hard to follow.

Could you break it up into multiple inline constants or something else more readable?

File: src/hooks/sessions/useAllowBookingActive.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import { Values } from 'SVConstants/values'
+
+const { CATEGORIES, SUB_CATEGORIES } = Values
+
+/**
+ * Pulls the state of the allow booking from the store and returns it
+ *
+ * @returns {boolean} - True if the allow booking is active
+ */
+export const useAllowBookingActive = () => {
```
Comment (mpcarolin) [inline src/hooks/sessions/useAllowBookingActive.js:11]: Yeah I think the "active" word choice isn't relevant.

**Review by mpcarolin — CHANGES_REQUESTED** [review]
Hey, code overall looks good, just have some minor requests.

I haven't tested it yet, but I will once I get back to my computer.

File: src/actions/session/booking/setAllowBookingActive.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * Then updates the store base on the result
+ * @param {Array.<import('SVModels/session').Session>} sessions
+ *
+ */
+export const setAllowBookingActive = sessions => {
+  const { items } = getStore()?.getState()
+  sessions = sessions || items?.sessions || noPropArr
+
+  const searchPath = [ 'allowBooking' ]
+  const allowBookingActive = sessions.some(session => get(session, searchPath))
```
Comment (lancetipton) [inline src/actions/session/booking/setAllowBookingActive.js:18]: The `get` method works with strings and arrays, so you could do this instead => 
```js
get(session, `allowBooking`)
```

File: src/actions/session/booking/setAllowBookingActive.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+const { CATEGORIES, SUB_CATEGORIES } = Values
+
+/**
+ * Checks if any sessions have allow booking set to true
+ * Then updates the store base on the result
+ * @param {Array.<import('SVModels/session').Session>} sessions
+ *
+ */
+export const setAllowBookingActive = sessions => {
```
Comment (lancetipton) [inline src/actions/session/booking/setAllowBookingActive.js:13]: I'm not crazy about the `setAllowBookingActive` name, it seems redundant. Why not just `setAllowBooking`
And really it's not setting it active. It could be setting it non-active based on the sessions

File: src/components/modals/filter.js [hunk 6]
```diff
@@ -193,21 +193,24 @@ const LabelButtons = React.memo(
...
 
   return useMemo(() => {
-    const { WAITING_LIST, ON_WAITING_LIST } = SESSION_BOOKING_STATES
+    const { WAITING_LIST, ON_WAITING_LIST, SELECT, SELECTED } = SESSION_BOOKING_STATES
     return reduceObj(
       bookingStates,
       (key, value, labels) => {
-        ;(waitingListActive ||
-          (value !== WAITING_LIST && value !== ON_WAITING_LIST)) &&
+        ;((waitingListActive &&
```
Comment (lancetipton) [inline src/components/modals/filter.js:203]: This seems pretty convoluted. I would break it up into smaller pieces.

**Review by lancetipton — CHANGES_REQUESTED** [review]
Overall looks good. Just a few minor things to look at.

File: src/actions/session/booking/setAllowBookingActive.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (subaya710) [reply src/actions/session/booking/setAllowBookingActive.js]: Great suggestion. Updates made as requested

File: src/components/modals/filter.js (hunk 1 above)
Comment (subaya710) [reply src/components/modals/filter.js]: Yeah , I wasn't sure about the double positive but went with it to maintain consistency. Updates made

File: src/components/modals/filter.js (hunk 2 above)
Comment (subaya710) [reply src/components/modals/filter.js]: Updates made . Hope this is a little clearer

File: src/hooks/sessions/useAllowBookingActive.js (hunk 3 above)
Comment (subaya710) [reply src/hooks/sessions/useAllowBookingActive.js]: Updates made

File: src/actions/session/booking/setAllowBookingActive.js (hunk 4 above)
Comment (subaya710) [reply src/actions/session/booking/setAllowBookingActive.js]: I liked Michael's option better so i used a direct session?.allowBooking . Hope thats fine

File: src/actions/session/booking/setAllowBookingActive.js (hunk 5 above)
Comment (subaya710) [reply src/actions/session/booking/setAllowBookingActive.js]: Updates made . Yeah i wasn't too sure about the double positive name either

File: src/components/modals/filter.js (hunk 6 above)
Comment (subaya710) [reply src/components/modals/filter.js]: Updates made . Hope this is a little clearer

**Review by mpcarolin — APPROVED** [review]
Looks good 👍

**Review by lancetipton — APPROVED** [review]
Looks good, thanks
